common/Services/PackageDeploymentService.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Corporation and Contributors
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using DevHome.Common.Exceptions;
+using Windows.Management.Deployment;
+
+namespace DevHome.Common.Services;
+
+public class PackageDeploymentService : IPackageDeploymentService
+{
+    private readonly PackageManager _packageManager = new ();
+
+    /// <inheritdoc />
+    public async Task RegisterPackageForCurrentUserAsync(string packageFamilyName, RegisterPackageOptions? options = null)
+    {
+        var result = await _packageManager.RegisterPackageByFamilyNameAsync(
+            packageFamilyName,
+            options?.DependencyPackageFamilyNames ?? new List<string>(),
+            options?.DeploymentOptions ?? DeploymentOptions.None,
+            options?.AppDataVolume,
+            options?.OptionalPackageFamilyNames ?? new List<string>());
+
+        if (!result.IsRegistered)
+        {
+            throw new RegisterPackageException(result.ErrorText, result.ExtendedErrorCode);
```

Comment:
Swallowing an exception is not ideal and using an exception to change code flow upstream isn't a good idea either.

Making an exception takes time.  I suggest to return Task<bool> and upstream code can determine what to do.  Use catches for exceptions that are unexpected.

Reply:
I’m not fully sure I understand the first sentence or where this is happening in the code in subject. Could you please elaborate?

In the suggested implementation in this PR, the caller of this (new) service method should handle the exception properly (and ideally log) in case a package failed to register. Catching `RegisterPackageException` will give the caller a readable error text and an inner exception `ExtendedErrorCode` both provided by the `RegisterPackageByFamilyNameAsync` method [return value](https://learn.microsoft.com/en-us/uwp/api/windows.management.deployment.deploymentresult?view=winrt-22621).

Reply:
Totes.  `RegisterAppInstallerAsync` calls `RegisterPackageForCurrentUserAsync`.  `RegisterPackageForCurrentUserAsync` throws a new exception if the package is not registered.

Then, `RegisterAppInstallerAsync` catches the exception and throws it away.  

Because you explicitly made a branch for "If App Installer isn't registered" it isn't an exception. Log the error message and signify to the caller that the package did not register.  Let the caller figure out how to handle the scenario.

Reply:
Discussed internally. Adding context here:

This is correct, `RegisterAppInstallerAsync` calls `RegisterPackageForCurrentUserAsync` and reduces the exception to a bool (registered / not registered). `RegisterPackageForCurrentUserAsync` is designed to be reusable service method, which in this case is being used to register AppInstaller.

The alternative suggested approach is, return [the original value](https://learn.microsoft.com/en-us/uwp/api/windows.management.deployment.deploymentresult?view=winrt-22621) from `RegisterPackageByFamilyNameAsync` and the caller will handle the logic. The challenge with this approach is that it complicates testing as `DeploymentResult` is not mockable in the unit tests. I added a comment for clarification.